private/bufpkg/bufconnect/interceptors.go

```diff
@@ -43,64 +44,120 @@ func NewSetCLIVersionInterceptor(version string) connect.UnaryInterceptorFunc {
-// which will look up an auth token by address and set it into the request header.  This is used for registry providers
-// where the token is looked up by the client address at the time of client construction (i.e. for clients where a
-// user is already authenticated and the token is stored in .netrc)
+// which will set the auth token into the request header by the provided option.
 //
 // Note that the interceptor returned from this provider is always applied LAST in the series of interceptors added to
 // a client.
-func NewAuthorizationInterceptorProvider(container app.EnvContainer) func(string) connect.UnaryInterceptorFunc {
+func NewAuthorizationInterceptorProvider(option SetAuthTokenOption) func(string) connect.UnaryInterceptorFunc {
 	return func(address string) connect.UnaryInterceptorFunc {
 		interceptor := func(next connect.UnaryFunc) connect.UnaryFunc {
 			return connect.UnaryFunc(func(
 				ctx context.Context,
 				req connect.AnyRequest,
 			) (connect.AnyResponse, error) {
-				envKey := tokenEnvKey
-				token := container.Env(envKey)
-				if token == "" {
-					envKey = ""
-					machine, err := netrc.GetMachineForName(container, address)
-					if err != nil {
-						return nil, fmt.Errorf("failed to read server password from netrc: %w", err)
-					}
-					if machine != nil {
-						token = machine.Password()
-					}
-				}
-				if token != "" {
-					req.Header().Set(AuthenticationHeader, AuthenticationTokenPrefix+token)
-				}
-				response, err := next(ctx, req)
-				if err != nil {
-					err = &ErrAuth{cause: err, tokenEnvKey: envKey}
-				}
-				return response, err
+				return option(req, ctx, next, address)
 			})
 		}
 		return interceptor
 	}
 }
 
-// NewAuthorizationInterceptorProviderWithToken returns a new provider function which, when invoked, returns an
-// interceptor which sets the provided auth token into the request header.  This is used for registry providers where
-// the token is known at provider creation (i.e. when logging in and explicitly pasting a token into stdin
-//
-// Note that the interceptor returned from this provider is always applied LAST in the series of interceptors added to
-// a client.
-func NewAuthorizationInterceptorProviderWithToken(token string) func(string) connect.UnaryInterceptorFunc {
-	return func(address string) connect.UnaryInterceptorFunc {
-		interceptor := func(next connect.UnaryFunc) connect.UnaryFunc {
-			return connect.UnaryFunc(func(
-				ctx context.Context,
-				req connect.AnyRequest,
-			) (connect.AnyResponse, error) {
-				if token != "" {
-					req.Header().Set(AuthenticationHeader, AuthenticationTokenPrefix+token)
-				}
-				return next(ctx, req)
-			})
+// SetAuthTokenOption is an option for NewAuthorizationInterceptorProvider
+type SetAuthTokenOption func(connect.AnyRequest, context.Context, connect.UnaryFunc, string) (connect.AnyResponse, error)
```

Comment:
I don't quite follow what the purpose of this is - why would we not just have two separate Interceptor providers?